pss/trojan/message.go

```diff
@@ -0,0 +1,215 @@
+	length  [2]byte // big-endian encoding of Message payload length
+	topic   Topic
+	payload []byte
+	padding []byte
+}
+
+// MaxPayloadSize is the maximum allowed payload size for the Message type, in bytes
+const MaxPayloadSize = 4030
+
+var hashFunc = storage.MakeHashFunc(storage.BMTHash)
+
+var errPayloadTooBig = fmt.Errorf("message payload size cannot be greater than %d bytes", MaxPayloadSize)
+var errEmptyTargets = errors.New("target list cannot be empty")
+var errVarLenTargets = errors.New("target list cannot have targets of different length")
+
+// NewTopic creates a new Topic variable with the given input string
+// the input string is taken as a byte slice and hashed
+func NewTopic(topic string) Topic {
+	return Topic(crypto.Keccak256Hash([]byte(topic)))
+}
+
+// newMessage creates a new Message variable with the given topic and payload
+// it finds a length and nonce for the message according to the given input and maximum payload size
+func newMessage(topic Topic, payload []byte) (Message, error) {
+	if len(payload) > MaxPayloadSize {
+		return Message{}, errPayloadTooBig
+	}
+
+	// get length as array of 2 bytes
+	payloadSize := uint16(len(payload))
+	lengthBuf := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuf, payloadSize)
+
+	// set random bytes as padding
+	paddingLen := MaxPayloadSize - payloadSize
+	padding := make([]byte, paddingLen)
+	if _, err := rand.Read(padding); err != nil {
+		return Message{}, err
+	}
+
+	// create new Message var and set fields
+	m := new(Message)
+	copy(m.length[:], lengthBuf[:])
+	m.payload = payload
+	m.padding = padding
+
+	return *m, nil
+}
+
+// Wrap creates a new trojan chunk for the given targets and Message
+// a trojan chunk is a content-addressed chunk made up of span, a nonce, and a payload which contains the Message
+// the chunk address will have a matching prefix with one of the targets
+func Wrap(targets [][]byte, m Message) (chunk.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	span := make([]byte, 8)
+	// 4064 bytes for trojan message + 32 bytes for nonce = 4096 bytes as payload for resulting chunk
+	binary.LittleEndian.PutUint64(span, chunk.DefaultSize)
+
+	chunk, err := toChunk(targets, span, m)
+	if err != nil {
+		return nil, err
+	}
+
+	return chunk, nil
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets [][]byte) error {
+	if len(targets) == 0 {
+		return errEmptyTargets
+	}
+	validLen := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLen {
+			return errVarLenTargets
+		}
+	}
+	return nil
+}
+
+// toChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iterating the BMT hash of the serialization of the trojan chunk fields until the desired nonce is found
```

Comment:
this is done by iteratively enumerating different nonces until the BMT hash of the serialization of the trojan chunk fields results in chunk address that has one of the targets as its prefix

Reply:
done